File: 剑指13.py
class Solution:
    def movingCount(self, m: int, n: int, k: int) -> int:
        def dfs(i, j):
            # 无效情形1：机器人走出界了
            if i >= m or j >= n:
                return 0
            #无效情形2：机器人走到一个之前访问过的（因为已经访问过了，所以对于结果‘能够到达多少格子’不能再算一次了）
            if (i, j) in visited:
                return 0
            #无效情形3：数位之和>k了
            sum_i = i // 10 + i % 10 #求出i的数位和（因为题目已经限制了0<=i<=99，所以只要除数为10即可）。下同
            sum_j = j // 10 + j % 10
            if sum_i + sum_j > k: #总的数位和
                return 0

            visited.add((i, j)) #每次的有效情形都需要把坐标记录到已访问数组内
            return 1 + dfs(i+1, j) + dfs(i, j+1) 

        visited = set()
        res = dfs(0, 0)
        return res

File: 剑指13.java
class Solution {
    int m, n;
    boolean[][] visited;
    public int movingCount(int m, int n, int k) {
        this.m = m;
        this.n = n;
        visited = new boolean[m][n];
        return dfs(0, 0, k);
    }

    private int dfs(int i, int j, int k){
        if(i>=m || j >= n || visited[i][j] == true || sumNum(i) + sumNum(j) > k){
            return 0;
        }
        visited[i][j] = true;
        return 1 + dfs(i+1, j, k) + dfs(i, j+1, k);
    }

    private int sumNum(int num){
        int ans = 0;
        while(num != 0){
            ans += num % 10;
            num = num / 10;
        }
        return ans;
    }
}
